### PR TITLE
fix: local dependencies correct bump from release to prerelease

### PR DESCRIPTION
**Pull request by davikawasaki**
Fixes bug #27. Tried to cover most of edge cases I've imagined, reusing code from semantic-release repository.

### CODE REVIEW

**Comment by antongolub** [comment]
@davikawasaki,

Wow! Outstanding improvement! But I need some time to sober up and be ready for review (10-12 hours).

CC @oljekechoro, @pismenskiy

**Comment by davikawasaki** [comment]
@antongolub np. Fixing the code complexity rn.

File: lib/utils.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+const { gt, lt, prerelease, rcompare } = require("semver");
+
+/**
+ * Get tag objects and convert them to a list of stringified versions.
+ * @param {array} tags Tags as object list.
+ * @returns {array} Tags as string list.
+ * @internal
+ */
+function tagsToVersions(tags) {
+	if (!tags) tags = [];
```
Comment (antongolub) [inline lib/utils.js:15]: How about "fail fast"?
```js
if (!tags) {
  return []
}
```

File: lib/utils.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Validates between two versions which is the highest.
+ * @param {string|undefined} version1 Version 1 to be compared with.
+ * @param {string|undefined} version2 Version 2 to be compared with.
+ * @param {integer} type (default as 1) Determines which type of verification is due (highest as 1, lowest as 0).
+ * @returns {string|undefined} Highest or lowest version.
+ * @internal
+ */
+function getHighestLowest(version1, version2, type = 1) {
```
Comment (antongolub) [inline lib/utils.js:27]: A bit confusing to me. Let's separate?
```js
const selectVersionBy = (predicate, version1, version2) => {

if (predicate && !version1 && !version2) {
   return predicate(version1, version2) ? version1 : version2
}

return version1 || version2
}

const getHighestVersion = selectVersionBy.bind(null, gt)
const getLowestVersion = selectVersionBy.bind(null, lt)
```

**Review by antongolub — CHANGES_REQUESTED** [review]

File: lib/updateDeps.js [hunk 3]
```diff
@@ -24,13 +26,43 @@ const getNextVersion = (pkg) => {
...
  */
-const getNextPreVersion = (pkg) => {
+const getNextPreVersion = (pkg, tags) => {
+	const tagFilters = [pkg._preRelease];
 	const lastVersion = pkg._lastRelease && pkg._lastRelease.version;
+	let version = `1.0.0-${pkg._preRelease}.1`;
 
-	return lastVersion ? semver.inc(lastVersion, "prerelease", pkg._preRelease) : `1.0.0-${pkg._preRelease}.1`;
+	if (pkg.name) tagFilters.push(pkg.name);
+	if (!tags || !tags.length) tags = getTags(pkg._branch || "master", { cwd: "." }, tagFilters);
```
[the remark this answers is not in the document]
Comment (davikawasaki) [reply lib/updateDeps.js:39]: No need. Removing it.

**Review by antongolub — APPROVED** [review]
@davikawasaki, 

LTGM. I'm going to merge. If everything is ok, once again I ask you to push these changes to the parent repo so that I don't have to "steal" your PR ))

**Comment by davikawasaki** [comment]
@antongolub great. I'll just add a PR refactor to reduce code complexity in a bit.

**Pull request merged**